File: test_projects/calculus/interactive_calculator.ts
import * as readline from 'readline';
import { parse } from './src/parser';
import { expressionUtils } from './src/utils';
import { Expression } from './src/expressionInterface';

interface Context {
    [variableName: string]: number;
}

class InteractiveCalculator {
    private rl: readline.Interface;

    constructor() {
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
    }

    private question(prompt: string): Promise<string> {
        return new Promise((resolve) => {
            this.rl.question(prompt, (answer) => {
                resolve(answer.trim());
            });
        });
    }

    private async showMenu(): Promise<string> {
        console.log('\n=== Mathematical Expression Calculator ===');
        console.log('1. Parse expression');
        console.log('2. Differentiate expression');
        console.log('3. Evaluate expression');
        console.log('4. Full workflow (parse → differentiate → evaluate)');
        console.log('5. Exit');
        console.log('==========================================');
        
        return await this.question('Choose an option (1-5): ');
    }

    private async parseExpression(): Promise<Expression | null> {
        try {
            const expressionStr = await this.question('Enter a mathematical expression: ');
            console.log(`\nParsing: "${expressionStr}"`);
            
            const parsedExpr = parse(expressionStr);
            console.log('✓ Successfully parsed!');
            console.log(`Expression structure: ${JSON.stringify(parsedExpr, null, 2)}`);
            console.log(`String representation: ${parsedExpr.toString()}`);
            
            return parsedExpr;
        } catch (error) {
            console.log(`✗ Parse error: ${error.message}`);
            return null;
        }
    }

    private async differentiateExpression(expr?: Expression): Promise<Expression | null> {
        try {
            let expression = expr;
            if (!expression) {
                const expressionStr = await this.question('Enter a mathematical expression: ');
                console.log(`\nParsing: "${expressionStr}"`);
                expression = parse(expressionStr);
                console.log('✓ Successfully parsed!');
            }

            const variable = await this.question('Enter variable to differentiate with respect to (e.g., x): ');
            console.log(`\nDifferentiating with respect to "${variable}"`);
            
            const derivative = expressionUtils.differentiate(expression, variable);
            console.log('✓ Successfully differentiated!');
            console.log(`Original: ${expression.toString()}`);
            console.log(`Derivative: ${derivative.toString()}`);
            
            return derivative;
        } catch (error) {
            console.log(`✗ Differentiation error: ${error.message}`);
            return null;
        }
    }

    private async evaluateExpression(expr?: Expression): Promise<number | null> {
        try {
            let expression = expr;
            if (!expression) {
                const expressionStr = await this.question('Enter a mathematical expression: ');
                console.log(`\nParsing: "${expressionStr}"`);
                expression = parse(expressionStr);
                console.log('✓ Successfully parsed!');
            }

            console.log(`\nExpression to evaluate: ${expression.toString()}`);
            
            // Extract variables from the expression
            const variables = this.extractVariables(expression);
            
            if (variables.length === 0) {
                console.log('No variables found. Evaluating directly...');
                const result = expressionUtils.evaluate(expression, {});
                console.log(`✓ Result: ${result}`);
                return result;
            }

            console.log(`Found variables: ${variables.join(', ')}`);
            const context: Context = {};
            
            for (const variable of variables) {
                const valueStr = await this.question(`Enter value for "${variable}": `);
                const value = parseFloat(valueStr);
                if (isNaN(value)) {
                    throw new Error(`Invalid number: "${valueStr}"`);
                }
                context[variable] = value;
            }
            
            const result = expressionUtils.evaluate(expression, context);
            console.log(`✓ Result: ${result}`);
            console.log(`Evaluation: ${expression.toString()} = ${result}`);
            console.log(`With values: ${Object.entries(context).map(([k, v]) => `${k}=${v}`).join(', ')}`);
            
            return result;
        } catch (error) {
            console.log(`✗ Evaluation error: ${error.message}`);
            return null;
        }
    }

    private extractVariables(expr: Expression): string[] {
        const variables = new Set<string>();
        
        const traverse = (node: Expression) => {
            if (node.type === 'variable') {
                variables.add((node as any).name);
            } else if (node.type === 'binaryOperation') {
                const binOp = node as any;
                traverse(binOp.left);
                traverse(binOp.right);
            } else if (node.type === 'unaryOperation') {
                const unOp = node as any;
                traverse(unOp.operand);
            } else if (node.type === 'functionCall') {
                const funcCall = node as any;
                funcCall.args.forEach(traverse);
            }
        };
        
        traverse(expr);
        return Array.from(variables).sort();
    }

    private async fullWorkflow(): Promise<void> {
        console.log('\n=== Full Workflow: Parse → Differentiate → Evaluate ===');
        
        // Step 1: Parse
        const expression = await this.parseExpression();
        if (!expression) return;
        
        // Step 2: Differentiate
        console.log('\n--- Differentiation Step ---');
        const derivative = await this.differentiateExpression(expression);
        if (!derivative) return;
        
        // Step 3: Evaluate original
        console.log('\n--- Evaluating Original Expression ---');
        await this.evaluateExpression(expression);
        
        // Step 4: Evaluate derivative
        console.log('\n--- Evaluating Derivative ---');
        await this.evaluateExpression(derivative);
    }

    private async showExamples(): Promise<void> {
        console.log('\n=== Expression Examples ===');
        console.log('Basic arithmetic: 2 + 3 * x');
        console.log('Powers: x^2 + 3*x + 1');
        console.log('Functions: sin(x) + cos(y)');
        console.log('Complex: ln(x^2 + 1) / (2*x + 3)');
        console.log('Trigonometry: sin(x) * cos(x) + tan(x/2)');
        console.log('Logarithms: log(10) + ln(e^x)');
        console.log('Nested: sin(cos(x)) + exp(ln(y))');
        console.log('============================\n');
    }

    async run(): Promise<void> {
        console.log('Welcome to the Interactive Mathematical Expression Calculator!');
        await this.showExamples();

        while (true) {
            try {
                const choice = await this.showMenu();
                
                switch (choice.trim()) {
                    case '1':
                        await this.parseExpression();
                        break;
                    case '2':
                        await this.differentiateExpression();
                        break;
                    case '3':
                        await this.evaluateExpression();
                        break;
                    case '4':
                        await this.fullWorkflow();
                        break;
                    case '5':
                        console.log('Goodbye! Thanks for using the calculator.');
                        this.rl.close();
                        return;
                    default:
                        console.log('Invalid choice. Please select 1-5.');
                }

                const continueChoice = await this.question('\nPress Enter to continue...');
            } catch (error) {
                console.log(`\n✗ Unexpected error: ${error.message}`);
                console.log('Please try again.');
            }
        }
    }
}

// Main execution
if (require.main === module) {
    const calculator = new InteractiveCalculator();
    calculator.run().catch(console.error);
} 

File: test_project_compare/opus_4_calculus/interactive_calculator.py
#!/usr/bin/env python3
"""
Interactive Mathematical Expression Calculator
A command-line interface for testing the opus 4 calculus implementation.
"""

import sys
import traceback
from utils import (
    evaluate, derivative, indefinite_integral, definite_integral,
    simplify, get_variables, substitute
)


class InteractiveCalculator:
    def __init__(self):
        self.last_expression = None
        self.last_result = None
    
    def show_menu(self):
        print('\n=== Mathematical Expression Calculator (Opus 4) ===')
        print('1. Parse and evaluate expression')
        print('2. Differentiate expression')
        print('3. Integrate indefinite')
        print('4. Integrate definite')
        print('5. Simplify expression')
        print('6. Get variables from expression')
        print('7. Substitute values in expression')
        print('8. Full workflow (evaluate → differentiate → integrate)')
        print('9. Show examples')
        print('0. Exit')
        print('======================================================')
        
        return input('Choose an option (0-9): ').strip()
    
    def get_input(self, prompt):
        """Get user input with proper handling."""
        try:
            return input(prompt).strip()
        except (EOFError, KeyboardInterrupt):
            print("\nGoodbye!")
            sys.exit(0)
    
    def parse_and_evaluate(self):
        """Parse and evaluate an expression."""
        print('\n--- Parse and Evaluate Expression ---')
        
        expression_str = self.get_input('Enter a mathematical expression: ')
        if not expression_str:
            print('✗ No expression provided.')
            return None
        
        self.last_expression = expression_str
        print(f'\nExpression: "{expression_str}"')
        
        try:
            # Get variables in the expression
            variables = get_variables(expression_str)
            
            if not variables:
                print('No variables found. Evaluating directly...')
                result = evaluate(expression_str)
                print(f'✓ Result: {result}')
                self.last_result = result
                return result
            
            print(f'Found variables: {", ".join(variables)}')
            
            # Get values for variables
            var_values = {}
            for var in variables:
                while True:
                    value_str = self.get_input(f'Enter value for "{var}" (or press Enter to skip): ')
                    if not value_str:
                        break
                    try:
                        var_values[var] = float(value_str)
                        break
                    except ValueError:
                        print(f'✗ Invalid number: "{value_str}". Please try again.')
            
            if var_values:
                result = evaluate(expression_str, var_values)
                print(f'✓ Result: {result}')
                print(f'With values: {", ".join(f"{k}={v}" for k, v in var_values.items())}')
                self.last_result = result
                return result
            else:
                result = evaluate(expression_str)
                print(f'✓ Symbolic result: {result}')
                self.last_result = result
                return result
                
        except Exception as e:
            print(f'✗ Error: {str(e)}')
            return None
    
    def differentiate_expression(self):
        """Differentiate an expression."""
        print('\n--- Differentiate Expression ---')
        
        expression_str = self.get_input('Enter expression (or press Enter to use last): ')
        if not expression_str and self.last_expression:
            expression_str = self.last_expression
            print(f'Using last expression: "{expression_str}"')
        elif not expression_str:
            print('✗ No expression provided.')
            return None
        
        variable = self.get_input('Enter variable to differentiate with respect to (e.g., x): ')
        if not variable:
            print('✗ No variable provided.')
            return None
        
        try:
            result = derivative(expression_str, variable)
            print(f'✓ Original: {expression_str}')
            print(f'✓ Derivative d/d{variable}: {result}')
            self.last_result = result
            return result
        except Exception as e:
            print(f'✗ Error: {str(e)}')
            return None
    
    def integrate_indefinite(self):
        """Compute indefinite integral."""
        print('\n--- Indefinite Integration ---')
        
        expression_str = self.get_input('Enter expression (or press Enter to use last): ')
        if not expression_str and self.last_expression:
            expression_str = self.last_expression
            print(f'Using last expression: "{expression_str}"')
        elif not expression_str:
            print('✗ No expression provided.')
            return None
        
        variable = self.get_input('Enter variable to integrate with respect to (e.g., x): ')
        if not variable:
            print('✗ No variable provided.')
            return None
        
        try:
            result = indefinite_integral(expression_str, variable)
            print(f'✓ Original: {expression_str}')
            print(f'✓ Indefinite integral ∫{expression_str} d{variable}: {result}')
            self.last_result = result
            return result
        except Exception as e:
            print(f'✗ Error: {str(e)}')
            print('This expression may require more advanced integration techniques.')
            return None
    
    def integrate_definite(self):
        """Compute definite integral."""
        print('\n--- Definite Integration ---')
        
        expression_str = self.get_input('Enter expression (or press Enter to use last): ')
        if not expression_str and self.last_expression:
            expression_str = self.last_expression
            print(f'Using last expression: "{expression_str}"')
        elif not expression_str:
            print('✗ No expression provided.')
            return None
        
        variable = self.get_input('Enter variable to integrate with respect to (e.g., x): ')
        if not variable:
            print('✗ No variable provided.')
            return None
        
        try:
            lower_str = self.get_input('Enter lower bound: ')
            upper_str = self.get_input('Enter upper bound: ')
            
            lower_bound = float(lower_str)
            upper_bound = float(upper_str)
            
            result = definite_integral(expression_str, variable, lower_bound, upper_bound)
            print(f'✓ Original: {expression_str}')
            print(f'✓ Definite integral from {lower_bound} to {upper_bound}: {result}')
            self.last_result = result
            return result
            
        except ValueError as e:
            print(f'✗ Invalid bounds: {str(e)}')
            return None
        except Exception as e:
            print(f'✗ Error: {str(e)}')
            return None
    
    def simplify_expression(self):
        """Simplify an expression."""
        print('\n--- Simplify Expression ---')
        
        expression_str = self.get_input('Enter expression (or press Enter to use last): ')
        if not expression_str and self.last_expression:
            expression_str = self.last_expression
            print(f'Using last expression: "{expression_str}"')
        elif not expression_str:
            print('✗ No expression provided.')
            return None
        
        try:
            result = simplify(expression_str)
            print(f'✓ Original: {expression_str}')
            print(f'✓ Simplified: {result}')
            self.last_result = result
            return result
        except Exception as e:
            print(f'✗ Error: {str(e)}')
            return None
    
    def get_variables_from_expression(self):
        """Get all variables in an expression."""
        print('\n--- Get Variables ---')
        
        expression_str = self.get_input('Enter expression (or press Enter to use last): ')
        if not expression_str and self.last_expression:
            expression_str = self.last_expression
            print(f'Using last expression: "{expression_str}"')
        elif not expression_str:
            print('✗ No expression provided.')
            return None
        
        try:
            variables = get_variables(expression_str)
            print(f'✓ Expression: {expression_str}')
            if variables:
                print(f'✓ Variables found: {", ".join(variables)}')
            else:
                print('✓ No variables found (expression contains only constants)')
            self.last_result = variables
            return variables
        except Exception as e:
            print(f'✗ Error: {str(e)}')
            return None
    
    def substitute_values(self):
        """Substitute values for variables in an expression."""
        print('\n--- Substitute Values ---')
        
        expression_str = self.get_input('Enter expression (or press Enter to use last): ')
        if not expression_str and self.last_expression:
            expression_str = self.last_expression
            print(f'Using last expression: "{expression_str}"')
        elif not expression_str:
            print('✗ No expression provided.')
            return None
        
        try:
            # Get variables first
            variables = get_variables(expression_str)
            if not variables:
                print('✓ No variables found in expression.')
                return expression_str
            
            print(f'Available variables: {", ".join(variables)}')
            
            substitutions = {}
            for var in variables:
                value_str = self.get_input(f'Enter value for "{var}" (or press Enter to skip): ')
                if value_str:
                    try:
                        substitutions[var] = float(value_str)
                    except ValueError:
                        print(f'✗ Invalid number for {var}: "{value_str}". Skipping.')
            
            if not substitutions:
                print('✓ No substitutions made.')
                return expression_str
            
            result = substitute(expression_str, substitutions)
            print(f'✓ Original: {expression_str}')
            print(f'✓ Substitutions: {", ".join(f"{k}={v}" for k, v in substitutions.items())}')
            print(f'✓ Result: {result}')
            self.last_result = result
            return result
            
        except Exception as e:
            print(f'✗ Error: {str(e)}')
            return None
    
    def full_workflow(self):
        """Demonstrate full workflow: evaluate → differentiate → integrate."""
        print('\n=== Full Workflow Demo ===')
        
        expression_str = self.get_input('Enter expression for full workflow: ')
        if not expression_str:
            print('✗ No expression provided.')
            return
        
        print(f'\nWorking with expression: "{expression_str}"')
        
        # Step 1: Parse and show variables
        print('\n1. Analysis:')
        try:
            variables = get_variables(expression_str)
            if variables:
                print(f'   Variables: {", ".join(variables)}')
            else:
                print('   No variables (constant expression)')
        except Exception as e:
            print(f'   ✗ Error getting variables: {e}')
        
        # Step 2: Simplify
        print('\n2. Simplification:')
        try:
            simplified = simplify(expression_str)
            print(f'   Simplified: {simplified}')
        except Exception as e:
            print(f'   ✗ Error simplifying: {e}')
        
        # Step 3: Differentiate
        print('\n3. Differentiation:')
        variable = self.get_input('   Enter variable to differentiate with respect to: ')
        if variable:
            try:
                deriv = derivative(expression_str, variable)
                print(f'   d/d{variable}: {deriv}')
            except Exception as e:
                print(f'   ✗ Error differentiating: {e}')
        
        # Step 4: Indefinite Integration
        print('\n4. Indefinite Integration:')
        if variable:
            try:
                indef = indefinite_integral(expression_str, variable)
                print(f'   ∫{expression_str} d{variable}: {indef}')
            except Exception as e:
                print(f'   ✗ Error integrating: {e}')
        
        # Step 5: Definite Integration
        print('\n5. Definite Integration:')
        if variable:
            try:
                lower_str = self.get_input('   Enter lower bound (or press Enter to skip): ')
                upper_str = self.get_input('   Enter upper bound (or press Enter to skip): ')
                
                if lower_str and upper_str:
                    lower = float(lower_str)
                    upper = float(upper_str)
                    definite = definite_integral(expression_str, variable, lower, upper)
                    print(f'   ∫[{lower},{upper}] {expression_str} d{variable}: {definite}')
                else:
                    print('   Skipping definite integration.')
            except Exception as e:
                print(f'   ✗ Error with definite integration: {e}')
        
        # Step 6: Evaluation
        print('\n6. Evaluation:')
        try:
            variables = get_variables(expression_str)
            if variables:
                print(f'   Variables found: {", ".join(variables)}')
                var_values = {}
                for var in variables:
                    value_str = self.get_input(f'   Enter value for {var} (or press Enter to skip): ')
                    if value_str:
                        try:
                            var_values[var] = float(value_str)
                        except ValueError:
                            print(f'   ✗ Invalid value for {var}')
                
                if var_values:
                    result = evaluate(expression_str, var_values)
                    print(f'   Result: {result}')
                else:
                    result = evaluate(expression_str)
                    print(f'   Symbolic result: {result}')
            else:
                result = evaluate(expression_str)
                print(f'   Result: {result}')
        except Exception as e:
            print(f'   ✗ Error evaluating: {e}')
        
        print('\n=== Workflow Complete ===')
    
    def show_examples(self):
        """Show example expressions and usage."""
        print('\n=== Expression Examples ===')
        print('Basic arithmetic: 2*x + 3, x^2 + 3*x + 1')
        print('Trigonometric: sin(x), cos(x), tan(x)')
        print('Logarithmic: ln(x), log(x) (base 10)')
        print('Exponential: exp(x), e^x')
        print('Mixed: x^2 + sin(x) + ln(x)')
        print('Complex: (x + 1) * (x - 1), sin(x)^2 + cos(x)^2')
        print('')
        print('Constants: e (Euler\'s number), pi')
        print('Variables: any letter(s), e.g., x, y, theta, var1')
        print('')
        print('Integration examples:')
        print('  - Polynomials: x^2, x^3 + 2*x')
        print('  - Trigonometric: sin(x), cos(x)')
        print('  - Exponential: exp(x), e^x')
        print('  - Logarithmic: 1/x → ln(x)')
        print('')
        print('Differentiation examples:')
        print('  - Power rule: x^n → n*x^(n-1)')
        print('  - Trig: sin(x) → cos(x), cos(x) → -sin(x)')
        print('  - Chain rule: sin(x^2) → cos(x^2)*2*x')
        print('=============================\n')
    
    def run(self):
        """Main calculator loop."""
        print('Welcome to the Opus 4 Mathematical Expression Calculator!')
        print('This calculator supports parsing, evaluation, differentiation, and integration.')
        
        self.show_examples()
        
        while True:
            try:
                choice = self.show_menu()
                
                if choice == '0':
                    print('Goodbye! Thanks for using the calculator.')
                    break
                elif choice == '1':
                    self.parse_and_evaluate()
                elif choice == '2':
                    self.differentiate_expression()
                elif choice == '3':
                    self.integrate_indefinite()
                elif choice == '4':
                    self.integrate_definite()
                elif choice == '5':
                    self.simplify_expression()
                elif choice == '6':
                    self.get_variables_from_expression()
                elif choice == '7':
                    self.substitute_values()
                elif choice == '8':
                    self.full_workflow()
                elif choice == '9':
                    self.show_examples()
                else:
                    print('Invalid choice. Please select 0-9.')
                
                if choice != '0':
                    input('\nPress Enter to continue...')
                    
            except KeyboardInterrupt:
                print('\n\nGoodbye!')
                break
            except Exception as e:
                print(f'\n✗ Unexpected error: {str(e)}')
                print('Please try again.')
                input('\nPress Enter to continue...')


if __name__ == '__main__':
    calculator = InteractiveCalculator()
    calculator.run() 